scheduler/src/cook/scheduler/constraints.clj

```diff
@@ -138,6 +148,49 @@
                                   (job->gpu-model-requested job-gpu-count-requested job (cached-queries/job->pool-name job)))]
     (->gpu-host-constraint job-gpu-count-requested job-gpu-model-requested)))
 
+; Cook uses mebibytes (MiB) for disk request and limit.
+; Convert bytes from k8s to MiB when passing to disk constraint,
+; and MiB back to bytes when submitting to k8s.
+(def disk-multiplier (* 1024 1024))
+
+(defrecord disk-host-constraint [job-disk-request job-disk-type]
+  JobConstraint
+  (job-constraint-name [this] (get-class-name this))
+  (job-constraint-evaluate
+    [this _ vm-attributes]
+    (job-constraint-evaluate this nil vm-attributes []))
+  (job-constraint-evaluate
+    [_ _ vm-attributes _]
+    (let [k8s-vm? (= (get vm-attributes "compute-cluster-type") "kubernetes")]
+      (if k8s-vm?
+        (let [vm-disk-type->space-available (get vm-attributes "disk")
+              ; k8s VMs all support disk, so they will only support jobs whose users requested disk or consume a default disk request
+              vm-satisfies-constraint? (if job-disk-request
+                                         (>= (get vm-disk-type->space-available job-disk-type 0) job-disk-request)
+                                         ; if job does not have a disk-request, pass true to always ignore the disk-host-constraint
+                                         true)]
+          [vm-satisfies-constraint? (when-not vm-satisfies-constraint?
+                                      "VM does not have enough disk space of requested disk type")])
+        ; Mesos jobs cannot request disk. If VM is a mesos VM, constraint always passes
+        [true]))))
+
+(defn build-disk-host-constraint
+  "Constructs a disk-host-constraint.
+  The constraint prevents a job from running on a host that does not have correct disk type that the job requested
+  and prevents a job from running on a host that does not have enough disk space. Use a constraint for disk binpacking instead of
+  using Fenzo because disk is not considered a first class resource in Fenzo."
+  [job]
+  (let [pool-name (cached-queries/job->pool-name job)]
+    ; If the pool does not have enable-constraint set to true, return nil
+    (when (regexp-tools/match-based-on-pool-name (config/disk) pool-name :enable-constraint? :default-value false)
+      (let [; If the user did not specify a disk request, use the default request amount for the pool
+            job-disk-request (or (-> job util/job-ent->resources :disk :request)
+                                 ; if disk config does not have default-request, use default-request of 10GiB
+                                 (regexp-tools/match-based-on-pool-name (config/disk) pool-name :default-request :default-value 10240))
+            job-disk-type (when job-disk-request
+                            (disk-type-requested (-> job util/job-ent->resources :disk :type) pool-name))]
+        (->disk-host-constraint job-disk-request job-disk-type)))))
```

Comment:
Would it make sense to wrap this in `(when job-disk-request ...)` so that we return `nil` from this function for jobs that have no explicit or defaulted disk request, avoiding creating the constraint altogether?

Reply:
this is already wrapped in the condition `(when enable-constraint?)`, and if `enable-constraint?` is true on a pool, that also means the job will have a disk-request (by default)

Reply:
updated comments, as discussed on Skype call